src/geojson_types.jl

```diff
@@ -0,0 +1,350 @@
+        names, types = property_schema(features)
+        return new{D,T}(bbox, features, crs, names, types)
+    end
+    FeatureCollection{D,T}(; bbox=nothing, features=Feature{D,T}[], crs=nothing) where {D,T} = FeatureCollection{D,T}(bbox, features, crs)
+end
+function FeatureCollection(; bbox=nothing, features::Vector{Feature{D,T}}, crs=nothing) where {D,T}
+    FeatureCollection{D,T}(bbox, features, crs)
+end
+
+
+features(fc::FeatureCollection) = getfield(fc, :features)
+Base.propertynames(fc::FeatureCollection) = getfield(fc, :names)
+
+function Base.getproperty(fc::FeatureCollection, name::Symbol)
+    if hasfield(typeof(fc), name)
+        getfield(fc, name)
+    else
+        getproperty.(fc, name)
+    end
+end
+
+function Base.iterate(fc::AbstractFeatureCollection{D,T}, state=1) where {D,T}
+    (1 <= state <= length(fc)) || return nothing
+    val = fc[state]::Feature{D,T}
+    return val, state + 1
+end
+
+Base.show(io::IO, fc::FeatureCollection) = print(io, "FeatureCollection with $(length(fc)) Features")
+Base.eltype(::Type{<:AbstractFeatureCollection{D,T}}) where {D,T} = Feature{D,T}
+Base.IteratorEltype(::Type{<:AbstractFeatureCollection}) = Base.HasEltype()
+Base.length(fc::AbstractFeatureCollection) = length(features(fc))
+Base.lastindex(fc::AbstractFeatureCollection) = length(features(fc))
+Base.IteratorSize(::Type{<:AbstractFeatureCollection}) = Base.HasLength()
+Base.size(fc::AbstractFeatureCollection) = size(features(fc))
+Base.getindex(fc::AbstractFeatureCollection, i::Int) = features(fc)[i]
+Base.IndexStyle(::AbstractFeatureCollection) = IndexLinear()
+
+bbox(x::GeoJSONT) = getfield(x, :bbox)
+
+"""
+    LazyFeatureCollection{D,T}(bbox::Union{Nothing,Vector{T}}, features::Vector{LazyFeature{D,T}}, crs::Union{Nothing,String})
+
+A FeatureCollection with `D` dimensional geometry in its features, but it's features are
+lazily parsed from the GeoJSON string. Indexing into the collection will parse the feature.
+This can be more efficient when interested in only a few features from a large collection,
+or parsing a very large collection iteratively without loading it all into memory.
+"""
+struct LazyFeatureCollection{D,T} <: AbstractFeatureCollection{D,T}
+    bbox::Union{Nothing,Vector{T}}
+    features::Vector{LazyFeature{D,T}}
+    crs::Union{Nothing,CRS}
+    # TODO Use features::JSON3.Array directly once we can parse the tape to find the
+    # feature offsets in the buffer?
+end
+features(fc::LazyFeatureCollection) = collect(fc.features)
+
+Base.show(io::IO, x::LazyFeatureCollection) = print(io, "LazyFeatureCollection with $(length(x.features)) features")
+Base.getindex(x::LazyFeatureCollection{D,T}, i::Int) where {D,T} = JSON3.read(codeunits(x.features[i]), Feature{D,T})::Feature{D,T}
+
+struct GeoJSONWrapper{D,T,X<:GeoJSONT{D,T}}
+    obj::X
+end
+GeoJSONWrapper{D,T}(obj::X) where {D,T,X<:GeoJSONT{D,T}} = GeoJSONWrapper{D,T,X}(obj)
+
+# symbol (from json string type) to struct mapping
+function geom_mapping(D, T)
```

Comment:
May need to make sure this specialises, you may want 
```suggestion
@inline function geom_mapping(D, ::Type{T}) where T
```

But I'm not sure about constant propagation for `D`. We should make sure this is type stable in real uses.

Reply:
julia> @time StructTypes.subtypes(GeoJSON.GeoJSONT{2,Float64})
  0.000001 seconds
(Point = GeoJSON.Point{2, Float64}, MultiPoint = GeoJSON.MultiPoint{2, Float64}, LineString = GeoJSON.LineString{2, Float64}, MultiLineString = GeoJSON.MultiLineString{2, Float64}, Polygon = GeoJSON.Polygon{2, Float64}, MultiPolygon = GeoJSON.MultiPolygon{2, Float64}, GeometryCollection = GeoJSON.GeometryCollection{2, Float64}, Feature = GeoJSON.Feature{2, Float64}, FeatureCollection = GeoJSON.FeatureCollection{2, Float64})

And @inferred works. What would you want to improve on this?

Reply:
Have you checked for a clean profile with ProfileView.jl? `@inferred` lies because it's just checking the output type is known, there can will be internal type stability.

Its not important I'm just going for worlds fastest GeoJSON reader now :smile:

Reply:
Squicky clean! There's probably improvements elsewhere, but we can improve them later without breaking anything.